libraries/chain_kv/include/b1/chain_kv/chain_kv.hpp

```diff
@@ -110,12 +111,141 @@ inline bytes get_next_prefix(const bytes& prefix) {
    return next_prefix;
 }
 
+namespace detail {
+   typedef __uint128_t uint128_t;
+
+   template<std::size_t Size, std::size_t N>
+   void copy_and_swap_to_buffer(char* to, const char* from) {
+      char* to_ptr = to + Size * (N - 1);
+      const char* from_ptr = from;
+      for (unsigned int i = 0; i < N; ++i) {
+         memcpy(to_ptr, from_ptr, Size);
+         to_ptr -= Size;
+         from_ptr += Size;
+      }
+   }
+
+   template<typename T>
+   struct value_storage;
+
+   template<typename T>
+   struct value_storage<T*> {
+      char* as_char_ptr(T* value) const { return reinterpret_cast<char*>(value); }
+      constexpr std::size_t size() const { return sizeof(T); }
+   };
+
+   template<typename T>
+   struct value_storage {
+      char* as_char_ptr(T& value) const { return reinterpret_cast<char*>(&value); }
```

Comment:
Similar question here, could these just be static methods?

I think the `as_char_ptr` can be constexpr too?